---
title: "Analysis"
author: "Julius J."
date: "`r Sys.Date()`"
geometry: "left=2cm,right=1cm,top=2cm,bottom=2cm"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/Documents/mieziai/calls/')
options(dplyr.summarise.inform = FALSE)

library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
```

Based on diagnostics, we use GoldenPromise v1 reference genome, duplicates removed (which does not make much difference).

The only genetic map I could find was on Morex x Barke, so we min-max rescale that to match GP chromosomes.

```{r readin}
# VCF fields:
# ID=MQ0F,Type=Float,Description="Fraction of MQ0 reads (smaller is better)"
# ID=MQ,Type=Integer,Description="Average mapping quality"
# ID=DP,Number=1,Type=Integer,Description="Raw read depth"
# ID=DP4,Number=4,Description="Number of high-quality ref-forward,
#   ref-reverse, alt-forward and alt-reverse bases
# MAF was calculated as DP4[3+4]/sum(DP4)

# load genetic map
# from  IBSC (2016-07-12): High-resolution GBS map of the Morex x Barke RIL population.
#  DOI:10.5447/ipk/2016/29
map = read.table("../refs/barley_morex_x_barke_high_resolution_gbs_map.tsv", h=T)
map = map[,2:4]
map$chr = as.numeric(substr(map$chr, 4, 4))

# Another genetic map from MIPS FTP
map2 = read.table("../refs/WGS_ANC.TXT", h=T)
map2$strain = sapply(strsplit(map2$sequence_id, "_"), "[[", 1)
map2 = filter(map2, strain=="morex")  # there are 3 identical strains there somehow

# snpEff annotations for the tw SNPs
ann = fread("../annots/annot_tw_mono.txt")
# already filtered by GT, MQ, DP
# filter strictly by MAF:
ann = filter(ann, MAF>0.9)

ann = mutate(ann, conseq = ifelse(grepl("HIGH", ANN), "high",
                           ifelse(grepl("MODERATE", ANN), "mod", "low"))) %>%
  mutate(conseq = factor(conseq, levels=c("high", "mod", "low")))

calls_out = tibble()
calls_out_cm = tibble()
calls_out_cm2 = tibble()
ann_out = tibble()
ann_out_cm = tibble()
for(chr in 1:7){
  callsA = bind_rows("AII"=fread(paste0("callsf_AII_HVgp_nodups_chr",chr,".txt")),
                     "tw"=fread(paste0("callsf_tw_HVgp_nodups_chr",chr,".txt")),
                    .id="pool")
  callsE = bind_rows("AII"=fread(paste0("exons_AII_HVgp_nodups_chr",chr,".txt")),
                    "tw"=fread(paste0("exons_tw_HVgp_nodups_chr",chr,".txt")),
                    .id="pool")
  calls = bind_rows("all"=callsA, "exon"=callsE, .id="type")
  
  # Filtering for mapping quality:
  calls = filter(calls, MQ>30)
  
  # Filtering by depth:
  calls = filter(calls, DP>=15, DP<90)  # about 10 % loss
  
  calls = calls[,c("pool", "type", "POS", "REF", "ALT", "DP", "MAF")]
  
  # Filtering only those present in both samples:
  callsBOTH = inner_join(filter(calls, pool=="AII", type=="all"),
                     filter(calls, pool=="tw", type=="all"),
                     by=c("POS", "REF", "ALT"),
                     suffix=c(".AII", ".tw")) %>%
    select(-one_of(c("pool.AII", "pool.tw")))

  # Keep only unique ones for either pool:
  callsUNIQ = anti_join(filter(calls, type=="all"),
                        callsBOTH, by=c("POS", "REF", "ALT"))

  # Filtering to remove roughly monomorphic markers:
  callsBOTH = filter(callsBOTH, MAF.AII<0.8 | MAF.tw<0.8)

  callsBOTH = pivot_longer(callsBOTH, c(DP.AII, DP.tw, MAF.AII, MAF.tw),
               names_to=c("stat", "pool"), names_sep="\\.", values_to="val") %>%
    spread("stat", "val")

  callsBOTH$type="shared"
  callsUNIQ$type="unique"
  calls = bind_rows(calls, callsBOTH, callsUNIQ)

  calls = mutate(calls, w = POS %/% 2e6 * 2)  # in 2 Mbp windows
  
  # merge in the genetic coords
  map_chr = map[map$chr==chr,]
  # basic rescaling, to at least match the length of GP chroms:
  range_old = range(map_chr$pos)
  range_new = range(calls$POS)
  map_chr$pos = (map_chr$pos-range_old[1]) / diff(range_old) * diff(range_new) + range_new[1]
  calls$cM = approx(map_chr$pos, map_chr$gbs_cM, calls$POS, rule=2)$y
  calls = mutate(calls, w_cM = cM %/% 1 * 1)  # in 1 cM windows
  
  # merge in the genetic coords, different map
  map2_chr = map2[map2$chr==chr,]
  range_old = range(map2_chr$Mb)
  map2_chr$pos = (map2_chr$Mb-range_old[1]) / diff(range_old) * diff(range_new) + range_new[1]
  calls$cM2 = approx(map2_chr$pos, map2_chr$cM, calls$POS, rule=2, ties=mean)$y
  calls = mutate(calls, w_cM2 = cM2 %/% 1 * 1)  # in 1 cM windows

  calls_sum = calls %>%
    group_by(w, pool, type) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  calls_sum_cm = calls %>%
    group_by(w_cM, pool, type) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  calls_sum_cm2 = calls %>%
    group_by(w_cM2, pool, type) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  
  # same for the tw snps w/ annotations
  ann_chr = ann[ann$CHR==chr,]
  calls_contr = filter(callsA, pool=="AII")
  ann_chr = bind_rows("shared"=semi_join(ann_chr, calls_contr, by=c("POS", "REF", "ALT")),
                      "unique"=anti_join(ann_chr, calls_contr, by=c("POS", "REF", "ALT")),
                      .id="shared")
  ann_chr$cM = approx(map_chr$pos, map_chr$gbs_cM, ann_chr$POS, rule=2)$y
  ann_chr = mutate(ann_chr, w = POS %/% 2e6*2, w_cM = cM %/% 1 * 1)  # in 2 Mbp / 1 cM windows
  ann_sum = ann_chr %>%
    group_by(w, conseq, shared) %>%
    summarize(n=n(), chr=chr)
  ann_sum_cm = ann_chr %>%
    group_by(w_cM, conseq, shared) %>%
    summarize(n=n(), chr=chr)
  

  calls_out = bind_rows(calls_out, calls_sum)
  calls_out_cm = bind_rows(calls_out_cm, calls_sum_cm)
  calls_out_cm2 = bind_rows(calls_out_cm2, calls_sum_cm2)
  ann_out = bind_rows(ann_out, ann_sum)
  ann_out_cm = bind_rows(ann_out_cm, ann_sum_cm)
}
```

# Coverage on physical map

```{r plotphys, echo=F, fig.width=9.5, fig.height=6.5}
# remember that depth changes basically only b/c of selection
# on called positions, and is quite flat overall
calls_out %>%
  ggplot(aes(x=w, y=medDP)) +
  geom_line(aes(col=pool), alpha=0.7) +
  facet_grid(type~chr, scales="free_x") +
  xlab("Mbp") + ylab("median depth") +
  theme_bw() + theme(legend.position="bottom")
ggplot(calls_out, aes(x=w, y=n)) +
  geom_line(aes(col=pool)) +
  facet_grid(type~chr, scales="free") +
  xlab("Mbp") + ylab("n calls") +
  theme_bw() + theme(legend.position="bottom")
ggplot(calls_out, aes(x=w, y=meanMAF)) +
  geom_line(aes(col=pool)) +
  facet_grid(type~chr, scales="free_x") +
  xlab("Mbp") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")
```

# Coverage on genetic map

```{r plotgen, echo=F, fig.width=9.5, fig.height=6.5}
ggplot(calls_out_cm, aes(x=w_cM, y=n)) +
  geom_line(aes(col=pool)) +
  facet_grid(type~chr, scales="free") +
  xlab("cM") + ylab("n calls") +
  theme_bw() + theme(legend.position="bottom")

ggplot(calls_out_cm, aes(x=w_cM, y=meanMAF, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(type~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")

# Frac AF>0.9, instead of mean AAF, cleans up the noise quite well
# as many positions have moderate AFs:
ggplot(calls_out_cm, aes(x=w_cM, y=mean11, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(type~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("frac. alt. AF>0.9") +
  theme_bw() + theme(legend.position="bottom")
```

# Coverage on a different genetic map

```{r plotgen2, echo=F, fig.width=9.5, fig.height=6.5}
ggplot(calls_out_cm2, aes(x=w_cM2, y=n)) +
  geom_line(aes(col=pool)) +
  facet_grid(type~chr, scales="free") +
  xlab("cM") + ylab("n calls") +
  theme_bw() + theme(legend.position="bottom")

ggplot(calls_out_cm2, aes(x=w_cM2, y=meanMAF, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(type~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")

# Frac AF>0.9, instead of mean AAF, cleans up the noise quite well
# as many positions have moderate AFs:
ggplot(calls_out_cm2, aes(x=w_cM2, y=mean11, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(type~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("frac. alt. AF>0.9") +
  theme_bw() + theme(legend.position="bottom")
```


# SNPs by annotated impact

```{r plotsnpeff, echo=F, fig.width=9.5, fig.height=6.5}
filter(ann_out, conseq=="low") %>%
  ggplot(aes(x=w, fill=conseq, y=n+1)) +
  geom_area(alpha=0.7) +
  geom_point(data=filter(ann_out, conseq!="low"), aes(col=conseq)) +
  facet_grid(shared~chr, scales="free_x") +
  scale_y_log10() +
  scale_fill_brewer(palette="OrRd", name="conseq", drop=F, direction=-1) + 
  scale_color_brewer(palette="OrRd", name="conseq", drop=F, direction=-1) + 
  xlab("Mbp") + ylab("number of calls") +
  theme_bw() + theme(legend.position="bottom")

filter(ann_out_cm, conseq=="low") %>%
  ggplot(aes(x=w_cM, fill=conseq, y=n+1)) +
  geom_area(alpha=0.7) +
  geom_point(data=filter(ann_out_cm, conseq!="low"), aes(col=conseq)) +
  facet_grid(shared~chr, scales="free_x") +
  scale_y_log10() +
  scale_fill_brewer(palette="OrRd", name="conseq", drop=F, direction=-1) + 
  scale_color_brewer(palette="OrRd", name="conseq", drop=F, direction=-1) + 
  xlab("cM") + ylab("number of calls") +
  theme_bw() + theme(legend.position="bottom")
```

# Comparing w/ individual runs shows correlated MAFs and DPs for the called variants

```{r indivcomp, fig.width=4, fig.height=4}
chr=1
cold = bind_rows("AII"=fread(paste0("../old/newcalls/callsf_AII-11_HVgp_chr",chr,".txt")),
                 "tw"=fread(paste0("../old/newcalls/callsf_tw-12_HVgp_chr",chr,".txt")),
                 .id="pool")
calls = bind_rows("AII"=fread(paste0("callsf_AII_HVgp_nodups_chr",chr,".txt")),
                 "tw"=fread(paste0("callsf_tw_HVgp_nodups_chr",chr,".txt")),
                 .id="pool")

tmp = inner_join(cold, calls, by=c("pool", "POS", "REF", "ALT"))
tmp = filter(tmp, POS<100e6, POS %% 10==0)
ggplot(tmp, aes(x=MAF.x, y=MAF.y, col=pool)) +
  geom_point(size=0.7) + theme_bw() +
  xlab("MAF indiv") + ylab("MAF pool")

filter(tmp, pool=="tw") %>%
  ggplot(aes(x=MAF.x, y=MAF.y)) +
  geom_point(size=0.7) + theme_bw() +
  xlab("MAF indiv. (tw)") + ylab("MAF pool (tw)")

filter(tmp, pool=="tw", DP.x<120, DP.y<150) %>%
  ggplot(aes(x=DP.x, y=DP.y)) +
  stat_smooth(method="lm", col="grey70") +
  geom_vline(xintercept=median(tmp$DP.x), col="grey50") +
  geom_hline(yintercept=median(tmp$DP.y), col="grey50") +
  geom_point(size=0.7) + theme_bw() +
  xlab("DP indiv. (tw)") + ylab("DP pool (tw)")

```

## However, correlation is much weaker after limiting to smaller DP
```{r mafcorr2}

tmp = inner_join(cold, calls, by=c("pool", "POS", "REF", "ALT")) %>%
  filter(POS<150e6, pool=="tw")
tmp2 = filter(tmp, DP.x<15, DP.x>5)
tmp2s = sample_n(tmp, nrow(tmp2))
bind_rows("5<DP<15"=tmp2, "random"=tmp2s, .id="variants") %>%
  ggplot(aes(x=MAF.x, y=MAF.y)) +
  facet_wrap(~variants) + 
  geom_point(size=0.7) + theme_bw() +
  xlab("MAF indiv. (tw)") + ylab("MAF pool (tw)")

```

# Replotting the mean MAF for small-DP variants

```{r readin2}
calls_out = tibble()
calls_out_cm = tibble()
for(chr in 1:7){
  calls = bind_rows("AII"=fread(paste0("callsf_AII_HVgp_nodups_chr",chr,".txt")),
                     "tw"=fread(paste0("callsf_tw_HVgp_nodups_chr",chr,".txt")),
                    .id="pool")
  contr_old = fread(paste0("../old/newcalls/callsf_tw-12_HVgp_chr",chr,".txt")) %>%
    filter(DP>10, DP<16)
  calls = semi_join(calls, contr_old, by=c("POS", "REF", "ALT"))
  
  # Filtering for mapping quality:
  calls = filter(calls, MQ>30)
  
  # Filtering by depth:
  calls = filter(calls, DP>=15, DP<90)  # about 10 % loss
  
  calls$type = "all"
  calls = calls[,c("pool", "type", "POS", "REF", "ALT", "DP", "MAF")]
  
  # Filtering only those present in both samples:
  callsBOTH = inner_join(filter(calls, pool=="AII", type=="all"),
                     filter(calls, pool=="tw", type=="all"),
                     by=c("POS", "REF", "ALT"),
                     suffix=c(".AII", ".tw")) %>%
    select(-one_of(c("pool.AII", "pool.tw")))

  # Keep only unique ones for either pool:
  callsUNIQ = anti_join(calls, callsBOTH, by=c("POS", "REF", "ALT"))

  # Filtering to remove roughly monomorphic markers:
  callsBOTH = filter(callsBOTH, MAF.AII<0.8 | MAF.tw<0.8)

  callsBOTH = pivot_longer(callsBOTH, c(DP.AII, DP.tw, MAF.AII, MAF.tw),
               names_to=c("stat", "pool"), names_sep="\\.", values_to="val") %>%
    spread("stat", "val")

  callsBOTH$type="shared"
  callsUNIQ$type="unique"
  calls = bind_rows(calls, callsBOTH, callsUNIQ)

  calls = mutate(calls, w = POS %/% 2e6 * 2)  # in 2 Mbp windows
  
  # merge in the genetic coords
  map_chr = map[map$chr==chr,]
  # basic rescaling, to at least match the length of GP chroms:
  range_old = range(map_chr$pos)
  range_new = range(calls$POS)
  map_chr$pos = (map_chr$pos-range_old[1]) / diff(range_old) * diff(range_new) + range_new[1]
  calls$cM = approx(map_chr$pos, map_chr$gbs_cM, calls$POS, rule=2)$y
  calls = mutate(calls, w_cM = cM %/% 1 * 1)  # in 1 cM windows
  
  calls_sum = calls %>%
    group_by(w, pool, type) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  calls_sum_cm = calls %>%
    group_by(w_cM, pool, type) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  
  calls_out = bind_rows(calls_out, calls_sum)
  calls_out_cm = bind_rows(calls_out_cm, calls_sum_cm)
}
```

```{r plotafterdplim, echo=F, fig.width=9.5, fig.height=5}
ggplot(calls_out, aes(x=w, y=n)) +
  geom_line(aes(col=pool)) +
  facet_grid(type~chr, scales="free") +
  xlab("Mbp") + ylab("n calls") +
  theme_bw() + theme(legend.position="bottom")
ggplot(calls_out, aes(x=w, y=meanMAF)) +
  geom_line(aes(col=pool)) +
  facet_grid(type~chr, scales="free_x") +
  xlab("Mbp") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")

ggplot(calls_out_cm, aes(x=w_cM, y=meanMAF, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(type~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")
ggplot(calls_out_cm, aes(x=w_cM, y=mean11, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(type~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("frac. alt. AF>0.9") +
  theme_bw() + theme(legend.position="bottom")
```


# Recalling on a consensus reference built from previous runs

```{r readin3}
cons_out = tibble()
cons_out_cm = tibble()
for(chr in 1:7){
  calls = bind_rows("AII"=fread(paste0("callsconsf_AII_HVgp_chr",chr,".txt")),
                     "tw"=fread(paste0("callsconsf_tw_HVgp_chr",chr,".txt")),
                    .id="pool")
  
  # Filtering for mapping quality:
  calls = filter(calls, MQ>30)
  
  # Filtering by depth:
  calls = filter(calls, DP>=15, DP<90)  # about 10 % loss
  
  calls = calls[,c("pool", "POS", "REF", "ALT", "DP", "MAF")]
  
  calls = mutate(calls, w = POS %/% 2e6 * 2)  # in 2 Mbp windows
  
  # merge in the genetic coords
  map_chr = map[map$chr==chr,]
  # basic rescaling, to at least match the length of GP chroms:
  range_old = range(map_chr$pos)
  range_new = range(calls$POS)
  map_chr$pos = (map_chr$pos-range_old[1]) / diff(range_old) * diff(range_new) + range_new[1]
  calls$cM = approx(map_chr$pos, map_chr$gbs_cM, calls$POS, rule=2)$y
  calls = mutate(calls, w_cM = cM %/% 1 * 1)  # in 1 cM windows
  
  calls_sum = calls %>%
    group_by(w, pool) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  calls_sum_cm = calls %>%
    group_by(w_cM, pool) %>%
    summarize(n=n(), mean11=mean(MAF>0.9), sum11=sum(MAF>0.9), medDP=median(DP),
              chr=chr, meanMAF=mean(MAF))
  
  calls_sum$refgen = "cons"
  calls_sum_cm$refgen = "cons"
  cons_out = bind_rows(cons_out, calls_sum)
  cons_out_cm = bind_rows(cons_out_cm, calls_sum_cm)
}

```

```{r plotcons, echo=F, fig.width=9.5, fig.height=4.5}
filter(calls_out, type=="all") %>%
  mutate(refgen="GP") %>%
  bind_rows(cons_out) %>%
  ggplot(aes(x=w, y=meanMAF)) +
  geom_line(aes(col=pool)) +
  facet_grid(refgen~chr, scales="free_x") +
  xlab("Mbp") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")

filter(calls_out_cm, type=="all") %>%
  mutate(refgen="GP") %>%
  bind_rows(cons_out_cm) %>%
  ggplot(aes(x=w_cM, y=meanMAF, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(refgen~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")
filter(calls_out_cm, type=="all") %>%
  mutate(refgen="GP") %>%
  bind_rows(cons_out_cm) %>%
  ggplot(aes(x=w_cM, y=mean11, col=pool)) +
  geom_line(alpha=0.4) +
  facet_grid(refgen~chr, scales="free_x") +
  geom_smooth(se=F, method="loess", formula="y~x") + 
  xlab("cM") + ylab("frac. alt. AF>0.9") +
  theme_bw() + theme(legend.position="bottom")
```

# Zooming in on the likely peak
```{r plotcons2, echo=F, fig.width=9.5, fig.height=5}
tmp_chr5 = filter(calls_out, type=="all", pool=="tw") %>%
  mutate(pool="tw_on_GP") %>%
  bind_rows(cons_out) %>%
  filter(chr==5)
pal = c("#DB8169",  "#619CFF", "#A7DCF2")
p1 = tmp_chr5 %>%
  ggplot(aes(x=w, y=meanMAF)) +
  geom_line(aes(col=pool)) +
  scale_color_manual(values=pal) + 
  xlab("Mbp") + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")
p2 = tmp_chr5 %>%
  ggplot(aes(x=w, y=meanMAF)) +
  geom_line(aes(col=pool)) +
  scale_color_manual(values=pal) +
  scale_x_continuous(n.breaks=10, expand=c(0,0)) +
  coord_cartesian(xlim=c(455, 540)) + 
  xlab(NULL) + ylab("mean alt. AF") +
  theme_bw() + theme(legend.position="bottom")

# high-impact, fixed in tw pool, not found in AII pool
ann_chr5 = filter(ann_out, chr==5, conseq %in% c("mod", "high"), shared=="unique")
p3 = ggplot(ann_chr5, aes(x=w, y=n, col=conseq)) +
  geom_point() +
  scale_color_manual(values=c("red", "orange")) +
  coord_cartesian(xlim=c(455, 540)) + 
  scale_x_continuous(n.breaks=10, expand=c(0,0)) +
  xlab(NULL) + ylab("n calls") +
  theme_bw() + theme(legend.position="bottom")

plot_grid(p1, plot_grid(p2, p3, nrow=2), ncol=2, rel_widths = c(5, 3))
```

## Chr. 5 mutations in tw are concentrated in this region

```{r ann_stats}
ann_chr5 = filter(ann, CHR==5)
frac.of.chr5 = 40e6/max(ann_chr5$POS)
frac.of.chr5

nchr5.twfixed = nrow(ann_chr5)
nchr5.twfixed
nrow(filter(ann_chr5, POS>490e6, POS<530e6)) / nchr5.twfixed

nchr5.twfixed.hi = nrow(filter(ann_chr5, conseq=="high" | conseq=="mod"))
nchr5.twfixed.hi
nrow(filter(ann_chr5, POS>490e6, POS<530e6, conseq=="high" | conseq=="mod")) / nchr5.twfixed.hi

contr_chr5 = fread("callsf_AII_HVgp_nodups_chr5.txt")
contr_chr5 = mutate(contr_chr5, reliable=DP>5 & MQ>30 & MQ0F<0.2 & MAF>0.05)
contr_chr5 = contr_chr5[,c("POS", "REF", "ALT", "reliable")]

ann_chr5 = left_join(ann_chr5, contr_chr5, by=c("POS", "REF", "ALT"))
nchr5.twfixed.uniq = nrow(filter(ann_chr5, is.na(reliable)))
nchr5.twfixed.uniq
nrow(filter(ann_chr5, is.na(reliable), POS>490e6, POS<530e6)) / nchr5.twfixed.uniq

ann_chr5 = filter(ann_chr5, POS>490e6, POS<530e6)
ann_chr5 = filter(ann_chr5, conseq!="low", is.na(reliable))
table(ann_chr5$conseq)
```

## High-impact mutations in the region

```{r hi_impact}
ann_chr5 = separate_rows(ann_chr5, ANN, sep=",") %>%
            separate(ANN,
                  c("Allele", "Effect", "Impact", "GeneName", "GeneID", "FeatureType", "FeatureID", 
                    "TranscrBiotype", "ExonRank", "HVGS.c", "HGVS.p", "pos.cDNA", "pos.CDS", "pos.PROT",
                    "Distance", "Errors"), sep="\\|")
ann_chr5 = select(ann_chr5, -one_of(c("GeneID", "FeatureType", "FeatureID", "TranscrBiotype",
                                  "Distance", "conseq", "reliable", "Allele")))

ann_hi = ann_chr5 %>% filter(Impact=="HIGH")
nrow(ann_hi)

# a few have very low DP4
filter(ann_hi, DP4>10) %>% nrow

# a few have some reference reads
filter(ann_hi, DP4>10, MAF==1) %>% nrow

# attach GO terms and save
biomart = fread("../refs/biomart_go_annots.txt")
biomart = unite(biomart, "Pfam.pos", "Pfam start":"Pfam end", sep="-") %>%
  unite("Pfam.domains", "Pfam ID":"Pfam.pos", sep=":")
biomart = group_by(biomart, `Gene stable ID`) %>%
  summarize(GO.term.name=paste(unique(`GO term name`), collapse="; "),
            Pfam.domains=paste(unique(`Pfam.domains`), collapse="; "))

ann_hi_go = biomart %>%
  inner_join(ann_hi, ., by=c("GeneName"="Gene stable ID"))

ann_mod = filter(ann_chr5, Impact=="MODERATE")
ann_mod_go = biomart %>%
  inner_join(ann_mod, ., by=c("GeneName"="Gene stable ID"))

print.data.frame(ann_hi_go)
write.table(ann_hi_go, "~/Documents/mieziai/calls/final_tw_chr5_hi.csv", sep="\t", quote=F, row.names=F)
write.table(ann_mod_go, "~/Documents/mieziai/calls/final_tw_chr5_mod.csv", sep="\t", quote=F, row.names=F)

# prepare gene position lists
gff = read.table("../refs/Hordeum_vulgare_goldenpromise.GPv1.54.gff3", h=F, sep="\t", quote="")
gff = filter(gff, V3=="gene")
gff$V9 = gsub(".*gene_id=(.*);.*", "\\1", gff$V9)
all(ann_chr5$GeneName %in% gff$V9)
gff_hi = filter(gff, V9 %in% ann_hi_go$GeneName)
gff_mod = filter(gff, V9 %in% ann_mod_go$GeneName)
# output positions for faidx
writeLines(paste(gff_hi$V1, paste(gff_hi$V4, gff_hi$V5, sep="-"), sep=":"),
           "genes/regions_hi.txt")
writeLines(paste(gff_mod$V1, paste(gff_mod$V4, gff_mod$V5, sep="-"), sep=":"),
           "genes/regions_mod.txt")

# output +-1000bp, on the consensus seq
gff_hi = mutate(gff_hi, V4=V4-1000, V5=V5+1000) %>%
  unite("pos", V4:V5, sep="-") %>%
  unite("region", c(V1, pos), sep=":") %>%
  arrange(region) %>%
  filter(!duplicated(region))
gff_mod = mutate(gff_mod, V4=V4-1000, V5=V5+1000) %>%
  unite("pos", V4:V5, sep="-") %>%
  unite("region", c(V1, pos), sep=":") %>%
  arrange(region) %>%
  filter(!duplicated(region))

writeLines(gff_hi$region,"genes/regions_buf1kb_hi.txt")
writeLines(gff_mod$region, "genes/regions_buf1kb_mod.txt")

```

